packages/components/src/navigation/use-create-navigation-tree.js

```diff
@@ -18,6 +23,47 @@ export const useCreateNavigationTree = () => {
 		removeNode: removeMenu,
 	} = useNavigationTreeNodes();
 
+	/**
+	 * Stores direct nested menus of menus
+	 * This makes it easy to traverse menu tree
+	 *
+	 * Key is the menu prop of the menu
+	 * Value is an array of menu keys
+	 */
+	const [ directSubMenus, setDirectSubMenus ] = useState( {} );
+	const getDirectSubMenus = ( menu ) => directSubMenus[ menu ] || [];
+
+	const traverseMenu = ( startMenu, callback ) => {
+		const visited = [];
+		let queue = [ startMenu ];
+		let current;
+
+		while ( queue.length > 0 ) {
+			current = getMenu( queue.shift() );
+
+			if ( ! current || visited.includes( current.menu ) ) {
+				continue;
+			}
+
+			visited.push( current.menu );
+			queue = [ ...queue, ...getDirectSubMenus( current.menu ) ];
+
+			callback( current );
+		}
+	};
+
+	const isMenuEmpty = ( menuToCheck ) => {
+		let count = 0;
+
```

Comment:
Correct me if I'm wrong, but wouldn't it be simpler to just do:

`return getMenu( menuToCheck ).isEmpty`

So long as we add every menu to `directSubMenus` when it is created?

That said, I'm guessing that the reason we have to add this algorithm is because submenus are not included in the `isEmpty` check? So a menu can be "empty" but also have submenus.

If that's the case, why not update the `isEmpty` boolean to consider submenus as well?

Reply:
> That said, I'm guessing that the reason we have to add this algorithm is because submenus are not included in the isEmpty check? So a menu can be "empty" but also have submenus.

^^^ This. Consider `Menu` -> `Sub Menu`. If `Menu` is empty, but `Sub Menu` isn't, then we still want to show `Menu`.

> If that's the case, why not update the isEmpty boolean to consider submenus as well?

🤔 How would that look like implementation-wise? How we determine `Sub Menu` is empty or not in the parent `Menu`? `isEmpty` is a prop that's manually set on the `NavigationMenu` component to indicate if it's empty or not. If we want to include sub-menus in this boolean as well, then we need to move all the logic into the parent menu.

Reply:
Hm, I see what you mean.

I think I understand this a bit better now. The reason we have this algorithm is that we consider a menu "empty" if all of its submenus and even grandchild submenus are empty. Therefore, we don't just need to check if a submeu exists for a given menu, but also that each submenu and its submeus is also empty.

I was thinking of this as in, we would consider a menu to be empty if it had 0 submenus, and it would not be empty if it had any submenus. But I see this isn't the case, and it makes a lot more sense now. 👍